### ISSUE TITLE
[GH Label Prediction] Extend GH Issue Summarization to Predict Labels

### ISSUE THREAD

**Issue by jlewi** [kubeflow/examples#56]
Should we extend our example on GH issue summarization to predict issue labels?

One of the points of the original [blog post](https://towardsdatascience.com/how-to-create-data-products-that-are-magical-using-sequence-to-sequence-models-703f86a231f8) was to train useful word embeddings on the entire corpus. So we could potentially use this to learn features that would then allow us to train models specific to repositories/orgs which likely have less data and their own taxonomy of labels.

Predicting issue labels would be useful for creating examples that highlight model analysis tooling. A lot of model analysis tools assume you can compute metrics like true positive/true negative etc... If we predict labels we can use actual labels to compute this.

With our existing text summarization example there's no obvious way to compute whether a summarization is accurate or not, which limits our ability to use it as an example of model analysis.

This would also be useful as a proxy for a large class of ecommerce problems where the goal is to dedupe related posts (e.g. different ebay postings for the same product).

@hamelsmu Any idea how difficult this would be and whether it would be valuable?

**Comment by hamelsmu** [comment]
We could do this, and it would be really cool.  Given the free-form and sparse nature of issue labels, Here is how I would approach it.  

1. Find the word-embeddings for the labels (from a pre-trained embedding like GloVe)
2. Pop the decoder off the Seq2Seq Model
3. Replace the decoder with a Linear Decoder that predicts the target-word embedding  (a regression problem).  The loss would be cosine_proximity in this case.  We can start with the pre-trained decoder from issue summarization and fine-tune it such that you get transfer learning effect.  
4.  Your model now predicts an embedding in the same vector-space as the word-embedding.  Now you can find issue-labels by nearest neighbor(s) lookup in the vocabulary.  This will even be able to find labels that are not in the training data.  

This is one idea of how to tackle it.  Another way is to take n top frequent labels and use that as pre-defined classes.

**Comment by hamelsmu** [comment]
cc: @rand @hohsiangwu

**Comment by rand** [comment]
This would be interesting, especially if we could get a finer grained view of whether something is a `bug` or a `fix` label, as those would likely have similar labels. If we go beyond the default label set to generic term/topics, we'd have to be careful to maintain a somewhat consistent set for a given project, otherwise the filtering utility of labels would be degraded. Also, we'd have to make the UX a bit smoother, so that people know that suggested labels have been introduced.

**Comment by jlewi** [comment]
This sounds fantastic; how can we help?

I think the authors of the [Tensor2Tensor](https://github.com/tensorflow/tensor2tensor) library are also eager to help. I think we might switch the example to use that library because

1. We think it will work better for distributed training
1. We think it will be easier to serve those models using TFServing.

**Comment by jlewi** [comment]
It looks like Tensor2Tensor can do text classification. We can follow the [imdb sentiment analysis example](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/data_generators/imdb.py). It looks like all we have to do is modify it to extract examples with the text and label.

Does the GitHub data have labels?

**Comment by hamelsmu** [comment]
Yes Github issues have labels associated with them.  Shall I pull the data (from a public source)?  What is the most effective way for me to help on this effort?

**Comment by jlewi** [comment]
If you could do one or both of the following things

1. Pull the data from a public source
2. Start writing code to train a model that would be fantastic.

If you ping me your Google account I will give you appropriate permissions in GCP project kubeflow-dev and you can use that (or your own project).

**Comment by hamelsmu** [comment]
@jlewi thanks, I emailed you my GCP id

**Comment by hamelsmu** [comment]
Ok I took a look at this it looks like T2T handles text summarization.  Maybe we could just use that instead of predicting issue labels (using the original data)?  I have a pretty dense back-log of things to do so this might take me a while to get to.  

In the mean-time I pulled some data that someone could use to predict issue labels for issues.  I used [GHArchive](https://www.gharchive.org/) to pull all issues from 2016-2018 that had a label.  If there is more than one label present, the labels will be separated by a comma.  Here is a screenshot of the data:

![image](https://user-images.githubusercontent.com/1483922/38789515-9f54515a-40ef-11e8-9925-448c686ce6b1.png)

There are approximately 1.7M Issues in this dataset.   This data can be downloaded here: 

https://storage.googleapis.com/hamel_githubissues/github_issues_with_labels_2016_2018.csv.zip
